plugins/processors/prometheusadapter/processor.go

```diff
@@ -5,24 +5,195 @@ package prometheusadapter
 
 import (
 	"context"
+	"os"
+	"strings"
 
+	"go.opentelemetry.io/collector/pdata/pcommon"
 	"go.opentelemetry.io/collector/pdata/pmetric"
 	"go.uber.org/zap"
+
+	"github.com/aws/amazon-cloudwatch-agent/plugins/processors/prometheusadapter/internal"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/prometheus"
+	"github.com/prometheus/common/model"
 )
 
 type prometheusAdapterProcessor struct {
 	*Config
-	logger *zap.Logger
+	logger          *zap.Logger
+	deltaCalculator *internal.DeltaCalculator
 }
 
 func newPrometheusAdapterProcessor(config *Config, logger *zap.Logger) *prometheusAdapterProcessor {
 	d := &prometheusAdapterProcessor{
-		Config: config,
-		logger: logger,
+		Config:          config,
+		logger:          logger,
+		deltaCalculator: internal.NewDeltaCalculator(),
 	}
 	return d
 }
 
 func (d *prometheusAdapterProcessor) processMetrics(_ context.Context, md pmetric.Metrics) (pmetric.Metrics, error) {
+
+	d.preprocessFilter(md)
+
+	md.ResourceMetrics().RemoveIf(func(rm pmetric.ResourceMetrics) bool {
+		rma := rm.Resource().Attributes()
+		sms := rm.ScopeMetrics()
+		sms.RemoveIf(func(sm pmetric.ScopeMetrics) bool {
+			metrics := sm.Metrics()
+			metrics.RemoveIf(func(m pmetric.Metric) bool {
+				return d.processMetric(m, rma)
+			})
+			return metrics.Len() == 0
+		})
+		return sms.Len() == 0
+	})
```

Comment:
Took a while to find it, but I remembered that there was a memory improvement done a while back that involved not creating references to the intermediate metric types https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/14729

```suggestion
	md.ResourceMetrics().RemoveIf(func(rm pmetric.ResourceMetrics) bool {
		rma := rm.Resource().Attributes()
		rm.ScopeMetrics().RemoveIf(func(sm pmetric.ScopeMetrics) bool {
			sm.Metrics().RemoveIf(func(m pmetric.Metric) bool {
				return d.processMetric(m, rma)
			})
			return sm.Metrics().Len() == 0
		})
		return rm.ScopeMetrics().Len() == 0
	})
```

Follows the pattern used by the filterprocessor as well https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/filterprocessor/metrics.go#L173

Reply:
wow, I'm surprised the compiler doesn't create the same output for both versions of the code. Looks better though, thanks